### PR TITLE
Add JsonnetVM methods for formatting

### PR DESCRIPTION
**Pull request by eagletmt**

### CODE REVIEW

File: ext/jsonnet/vm.c [hunk 1]
```diff
@@ -255,6 +257,112 @@ vm_set_max_trace(VALUE self, VALUE val)
...
+static VALUE
+vm_set_fmt_sort_imports(VALUE self, VALUE val)
+{
+    struct jsonnet_vm_wrap *vm = rubyjsonnet_obj_to_vm(self);
+    jsonnet_fmt_sort_imports(vm->vm, RTEST(val) ? 1 : 0);
+    return val;
+}
+
+static VALUE
+vm_set_fmt_file(VALUE self, VALUE fname, VALUE encoding)
```
Comment (yugui) [inline ext/jsonnet/vm.c:335]:
```suggestion
vm_fmt_file(VALUE self, VALUE fname, VALUE encoding)
```

File: ext/jsonnet/vm.c [hunk 2]
```diff
@@ -255,6 +257,112 @@ vm_set_max_trace(VALUE self, VALUE val)
...
+    FilePathValue(fname);
+    result = jsonnet_fmt_file(vm->vm, StringValueCStr(fname), &error);
+    if (error) {
+      raise_format_error(vm->vm, result, rb_enc_get(fname));
+    }
+    return str_new_json(vm->vm, result, enc);
+}
+
+static VALUE
+vm_set_fmt_snippet(VALUE self, VALUE snippet, VALUE fname)
```
Comment (yugui) [inline ext/jsonnet/vm.c:351]:
```suggestion
vm_fmt_snippet(VALUE self, VALUE snippet, VALUE fname)
```

File: ext/jsonnet/vm.c [hunk 3]
```diff
@@ -272,22 +380,25 @@ rubyjsonnet_init_vm(VALUE mJsonnet)
...
+    rb_define_method(cVM, "fmt_indent=", vm_set_fmt_indent, 1);
+    rb_define_method(cVM, "fmt_max_blank_lines=", vm_set_fmt_max_blank_lines, 1);
+    rb_define_method(cVM, "fmt_string=", vm_set_fmt_string, 1);
+    rb_define_method(cVM, "fmt_comment=", vm_set_fmt_comment, 1);
+    rb_define_method(cVM, "fmt_pad_arrays=", vm_set_fmt_pad_arrays, 1);
+    rb_define_method(cVM, "fmt_pad_objects=", vm_set_fmt_pad_objects, 1);
+    rb_define_method(cVM, "fmt_pretty_field_names=", vm_set_fmt_pretty_field_names, 1);
+    rb_define_method(cVM, "fmt_sort_imports=", vm_set_fmt_sort_imports, 1);
+    rb_define_method(cVM, "fmt_file", vm_set_fmt_file, 2);
+    rb_define_method(cVM, "fmt_snippet", vm_set_fmt_snippet, 2);
```
Comment (yugui) [inline ext/jsonnet/vm.c:392]:
```suggestion
    rb_define_method(cVM, "fmt_file", vm_fmt_file, 2);
    rb_define_method(cVM, "fmt_snippet", vm_fmt_snippet, 2);
```

File: ext/jsonnet/vm.c [hunk 4]
```diff
@@ -255,6 +257,112 @@ vm_set_max_trace(VALUE self, VALUE val)
...
+vm_set_fmt_string(VALUE self, VALUE val)
+{
+    char *str;
+    struct jsonnet_vm_wrap *vm = rubyjsonnet_obj_to_vm(self);
+    str = StringValueCStr(val);
+    if (str[0] == '\0' || str[1] != '\0' || (str[0] != 'd' && str[0] != 's' && str[0] != 'l')) {
+      rb_raise(rb_eArgError, "fmt_string only accepts 'd', 's', or 'l'");
+    }
+    jsonnet_fmt_string(vm->vm, str[0]);
+    return val;
```
Comment (yugui) [inline ext/jsonnet/vm.c:286]: I suspect if this is more straightforward and slightly faster.

```suggestion
    struct jsonnet_vm_wrap *vm = rubyjsonnet_obj_to_vm(self);
    const char *ptr;
    if (RSTRING_LEN(str) != 1) {
        rb_raise(rb_eArgError, "fmt_string is too long");
    }
    ptr = RSTRING_PTR(str);
    switch (*ptr) {
    case 'd':
    case 's':
    case 'l':
        jsonnet_fmt_string(vm->vm, str[0]);
        return val;
    default:
        rb_raise(rb_eArgError, "fmt_string only accepts 'd', 's', or 'l'");
    }
```

File: ext/jsonnet/vm.c [hunk 5]
```diff
@@ -255,6 +257,112 @@ vm_set_max_trace(VALUE self, VALUE val)
...
+}
+
+static VALUE
+vm_set_fmt_comment(VALUE self, VALUE val)
+{
+    char *str;
+    struct jsonnet_vm_wrap *vm = rubyjsonnet_obj_to_vm(self);
+    str = StringValueCStr(val);
+    if (str[0] == '\0' || str[1] != '\0' || (str[0] != 'h' && str[0] != 's' && str[0] != 'l')) {
```
Comment (yugui) [inline ext/jsonnet/vm.c:295]: ditto.

File: ext/jsonnet/vm.c [hunk 6]
```diff
@@ -255,6 +257,112 @@ vm_set_max_trace(VALUE self, VALUE val)
...
+    return val;
+}
+
+static VALUE
+vm_set_fmt_string(VALUE self, VALUE val)
+{
+    char *str;
+    struct jsonnet_vm_wrap *vm = rubyjsonnet_obj_to_vm(self);
+    str = StringValueCStr(val);
+    if (str[0] == '\0' || str[1] != '\0' || (str[0] != 'd' && str[0] != 's' && str[0] != 'l')) {
```
Comment (yugui) [inline ext/jsonnet/vm.c:282]: Should define Ruby-level constants for these values

File: ext/jsonnet/vm.c (hunk 4 above)
Comment (eagletmt) [reply ext/jsonnet/vm.c]: I think `StringValue(str)` is required before RSTRING_LEN/RSTRING_PTR to ensure str is String. So I've committed this change https://github.com/yugui/ruby-jsonnet/pull/17/commits/9bab98fa474ad7f76ae99f3c42c5b585b53cd566

File: ext/jsonnet/vm.c (hunk 6 above)
Comment (eagletmt) [reply ext/jsonnet/vm.c]: Like this? https://github.com/yugui/ruby-jsonnet/pull/17/commits/e47a1a3c75ac8a625e671ffb2dcf43d2ed3717cb

File: ext/jsonnet/vm.c [hunk 7]
```diff
@@ -272,22 +396,32 @@ rubyjsonnet_init_vm(VALUE mJsonnet)
...
+    rb_define_method(cVM, "fmt_indent=", vm_set_fmt_indent, 1);
+    rb_define_method(cVM, "fmt_max_blank_lines=", vm_set_fmt_max_blank_lines, 1);
+    rb_define_method(cVM, "fmt_string=", vm_set_fmt_string, 1);
+    rb_define_method(cVM, "fmt_comment=", vm_set_fmt_comment, 1);
+    rb_define_method(cVM, "fmt_pad_arrays=", vm_set_fmt_pad_arrays, 1);
+    rb_define_method(cVM, "fmt_pad_objects=", vm_set_fmt_pad_objects, 1);
+    rb_define_method(cVM, "fmt_pretty_field_names=", vm_set_fmt_pretty_field_names, 1);
+    rb_define_method(cVM, "fmt_sort_imports=", vm_set_fmt_sort_imports, 1);
+    rb_define_method(cVM, "fmt_file", vm_fmt_file, 2);
+    rb_define_method(cVM, "fmt_snippet", vm_fmt_snippet, 2);
```
Comment (yugui) [inline ext/jsonnet/vm.c:408]: Sorry. I overlooked this point at the first comment.

As long as you intend to call these methods from their wrappers, these methods should be private.
Should use `rb_define_private_method` as `eval_file` or `eval_snippet` does.

File: ext/jsonnet/vm.c (hunk 4 above)
Comment (yugui) [reply ext/jsonnet/vm.c]: good point. thank you.

**Comment by eagletmt** [comment]
Oh, this PR is broken by updating jsonnet to v0.16.
https://travis-ci.org/github/yugui/ruby-jsonnet/builds/704648423
I'm making a patch to fix the crash.

**Pull request merged**